### ISSUE TITLE
Default dart Map behaviour seems quadratic for some key distributions

### ISSUE THREAD

**Issue by tarobins** [dart-lang/sdk#40331]
This tracker is for issues related to:
* Dart VM

* Dart SDK Version (`dart --version`)
Dart VM version: 2.8.0-dev.5.0 (Thu Jan 23 10:31:03 2020 +0100) on "macos_x64"

Missing some or all of the above might make the issue take longer or be
impossible to act on.

Depending on how keys are constructed, the default map seems to have a runtime complexity that is approximately quadratic 

Using the code in my [map_timer repo](https://github.com/tarobins/map_timer).  I tested the insertion of a sequence of pow(2, i) keys for 0 <= i <= 19, where the keys for each sequence were of the form:

1. j, j = 0..i or
1. j << 40, j = 0..i

for case 1, the growth of the runtime seemed to be linear, as seen in the timings gathered here https://docs.google.com/spreadsheets/d/1PA6K8B3j_dX9_YGaya-ddUdGwgVXFTP6_QOkt6BluQg/edit#gid=0&range=A1

for case 2, the growth seemed to be quadratic, as seen in the timings gathered here 
https://docs.google.com/spreadsheets/d/1PA6K8B3j_dX9_YGaya-ddUdGwgVXFTP6_QOkt6BluQg/edit#gid=1547091203&range=A1.

**Comment by tatumizer** [comment]
The problem is not related to map or hashCode. Compiler fails to expand the value to 64 bits: `print(1<<40);` prints 0;

**Comment by tarobins** [comment]
I'm getting 1099511627776

```
void main(List<String> arguments) {
  print(1 << 40);
}
```

/usr/local/opt/dart/libexec/bin/dart --enable-asserts --enable-vm-service:58095 /Users/robinsontom/scratch/map_timer/bin/main.dart
Observatory listening on http://127.0.0.1:58095/_Rid_-G7n3g=/

1099511627776

Process finished with exit code 0

**Comment by tarobins** [comment]
I do get 0 in dartpad.

**Comment by tatumizer** [comment]
Yes, I'm running on dartpad. You have to check what hashCode is for all values you compute. The result can be 64-bit, but hashCode may take just lower 32 bits, so you get 0.
`print((1<<40).hashCode);`

**Comment by jonahwilliams** [comment]
The behavior on dartpad is expected, since that executes Dart compiled to JavaScript (which only supports 32bit ints)

**Comment by tatumizer** [comment]
Sure, my mistake. I finally made it to my comp and ran it on 64-bit dart.
The problem is that hashCode for an int is equal to the int.
E.g.
```
print((2<<40).hashCode);
print((2<<40));
```
prints
```
1099511627776
1099511627776
```
if the number of buckets in the map is (hypothetically) a power of 2, then all values `i<<40` will go to the same bucket. Normally, hashCode would scramble the bits of int to get a hashCode. This doesn't happen.

**Comment by tarobins** [comment]
yes, we need to figure out a way to take the higher-order bits into account while not slowing down cases where the behaviour is currently good.  

In some earlier experimentation, I use flutter's [hashValues function](https://api.flutter.dev/flutter/dart-ui/hashValues.html) to mash up the bits from `Color.hashCode` (e.g. using `hashValues(Color.hashCode, 0)`), which is just the int representation of the ARGB value, and doing so improved the performance of Map on Color keys; however, when hashing on smaller integers, e.g. the individual components of the Colors, using hashValues actually slowed things down. 

We also need to figure out where to implement a solution: do we modify how classes like the default implementation of Map use hashCode, modify the hashCode function on some of the existing Dart classes, such as Int and Color, or some combo of the two.

**Comment by tatumizer** [comment]
I vaguely remember dart 1 having a much better hash function for ints (some google's standard algo was used back then if I'm not mistaken). Not sure what happened with that implementation.

**Comment by tarobins** [comment]
One possibility would be to have a parameter to `hashCode` that specifies the number of bits, n, in the hash code that will be used, so the class implementing hashCode could return a hash code where the lower n bits contain useful data.  Obviously a big change.  

Alternatively, could set up the assumption that collections always will use some minimum number of bits, m, when bucketing items, so the class implementing hashCode can make sure that the m least significant bits vary appropriately.  The collections could use more bits from the hash, but will at least use the least significant m.

**Comment by davidmorgan** [comment]
A bit more context/background: Dart Map in the VM does not currently do any additional processing of the hashCode to make sure higher bits are used. As an additional data point, the Java SDK implementation does precisely that: a simple+fast mangling of the bits to make it much harder to hit the worst case seen here.

**Comment by tarobins** [comment]
So we could just do something like what java does.  My concern is that java ints are 32 bits and dart's are 64, so if we just fold the bits over once, we could still easily run into bad cases where the bits that vary in the keys are still not used in the bucketing.

**Comment by tatumizer** [comment]
Just curious: what if you call Flutter's hashValues with 2 parts: high-order bits and low-order bits:
`hashValues(i>>32, i&0xFFFFFFFF)`
It may mix things up pretty well.
